optaplanner-core/src/main/java/org/optaplanner/core/api/score/holder/AbstractScoreHolder.java

```diff
@@ -67,16 +68,44 @@ public boolean isConstraintMatchEnabled() {
     // Worker methods
     // ************************************************************************
 
+    protected class MultiLevelConstraintUndoListener implements ActivationUnMatchListener {
+
+        private Map<Integer, ActivationUnMatchListener> activationUnMatchListenerMap;
+
+        public MultiLevelConstraintUndoListener(int scoreLevel, ActivationUnMatchListener activationUnMatchListener) {
+            this.activationUnMatchListenerMap = new HashMap<Integer, ActivationUnMatchListener>();
```

Comment:
Doing some experiments with visualvm. I 've got 2 checkouts and 2 intellij windows of optaplanner (one with these changes added and one which I call old-optaplanner because it's without those changes) - running them next to each other to see difference.